### PR TITLE
Fix memleak in pouchdb-core

### LINKED ISSUE #4868: [nodejs] memleak when creating PouchDB instance
I am seeing the issue in production with slightly more complicated code (the memleak is obvious after 30 minutes but the databases are all different and using the `http` adapter, etc.), but here is a simple example that also exhibits the issue:

```
var PouchDB, debug, dgram, heapUsed, socket;
dgram = require('dgram');
PouchDB = require('pouchdb');
debug = (require('debug'))('test');
heapUsed = null;
socket = dgram.createSocket('udp4');
socket.on('message', function(msg, rinfo) {
  var db;
  debug("Received " + msg.length + " bytes from " + rinfo.address + ":" + rinfo.port);
  db = new PouchDB('goodluck');
  return db.info(function(info) {
    var last_heapUsed, memory;
    memory = process.memoryUsage();
    last_heapUsed = heapUsed;
    heapUsed = memory.heapUsed;
    return debug('info', last_heapUsed != null ? heapUsed - last_heapUsed : null);
  });
});
socket.bind(7124);
```

I then run `while true; do echo ping | nc -u -w 1 127.0.0.1 7124; done` in a shell to trigger the `message` handler.

Forcing a `gc()` call for each message (using `--expose_gc --expose_gc_as=gc`):
- There is no leak if I remove the `new PouchDB...` code. (The difference is '0' after about 6-7 iterations.)
- The leak is about the same size with or without the call to `info()`.
- The leak is about 2 to 3ko with nodejs 5.4.1 or 5.5.0, about 10 to 14ko with nodejs 4.2.1 or 4.2.6.

I'm obviously not sure this might be a PouchDB issue, but I'm at a loss to analyze where the issue might come from (I've done some runs with [`heapdump`](https://www.npmjs.com/package/heapdump) but it apparently doesn't work when `--expose_gc` is active).

### PR DESCRIPTION
**Pull request by shimaore**
This partially fixes #4868 by removing existing memleak in pouchdb-core.
This is a revised and expanded version of a previous PR.
This PR contains:
- the memleak testing code (which is otherwise available in [my memleak-test branch](https://github.com/shimaore/pouchdb/tree/memleak-test)), with self-test and pouchdb-core tests enabled;
- changes to gitignore and package.json to enable the memleak testing code;
- changes to pouchdb-core to remove the memleak on `close()` and `destroy()`. This changeset is different from the one I previously submitted and performs better (i.e. doesn't leak on `close()` as far as the test suite can tell).

There are other memleaks in (at least) the http adapter (still tracking this one, I suspect the leak is in a dependency) and leveldown (I noticed at least one static `LazySet` in the code which might partially explain it, since `LazySet` doesn't have a method to remove entries).

### CODE REVIEW

**Comment by shimaore** [comment]
Found a couple more locations for corrections, based on https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#32-leaking-arguments, will add them shortly.

File: tests/memleak/test.memleak.js [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+      var db = new PouchDB(db_name,opts);
+      function Finally() { return db.close(); }
+      db_name = null;
+      return db.info()
+      .then( Finally, Finally )
+      .then(function () {
+        db = null;
+        return sleep(20);
+      });
+    };
```
Comment (daleharvey) [inline tests/memleak/test.memleak.js]: Need to remove this semi colon, failing eslint

File: packages/node_modules/pouchdb-adapter-indexeddb/src/index.js [hunk 2]
```diff
@@ -33,7 +33,10 @@ function IdbPouch(dbOpts, callback) {
   // the database handle as the first argument
   var $ = function (fun) {
     return function () {
-      var args = Array.prototype.slice.call(arguments);
+      var args = new Array(arguments.length);
+      for(var i = 0; i < args.length; ++i) {
+        args[i] = arguments[i];
+      }
```
Comment (daleharvey) [inline packages/node_modules/pouchdb-adapter-indexeddb/src/index.js]: Whats the reason for doing this? is it part of the fix for memleak?

File: tests/memleak/test.memleak.js [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+    api.info = function DummyPouchAdapterInfo(callback) {
+      if(callback) {
+        callback(null,{dummy:true});
+      } else {
+        return Promise.resolve({dummy:true});
+      }
+    };
+
+    if(callback) {
```
Comment (daleharvey) [inline tests/memleak/test.memleak.js]: I need to put a fix in for eslint, but in general we prefer to have a space before parens

    if (callback) 
    api.info = function DummyPouchAdapterInfo (callback) {

etc, not a huge deal if you dont fix, will probably just do a batch fix when we fix eslint

**Comment by daleharvey** [comment]
Overall this look great, the tests are amazing, really like seeing the comparitive tests to clean state vs broken, theres a few minor comments in there, if we could avoid the `var args = Array.prototype.slice.call(arguments);` change if possible and do that seperately it would be good, unless its actually involved in the memory leak, needs the semi colon fix for the tests though

**Comment by shimaore** [comment]
@daleharvey I've reverted the changes regarding `arguments`: I had noted in the commits that they addressed memleaks in pouchdb-http-adapter but I can't reproduce those changes anymore (but I'm now testing with Node.js 7 so there might be a difference there as well — anyway).

I also checked eslint and inserted spaces after `if`, but eslint complained about `Unexpected space before function parentheses` when I tried `function ABC ()` so I left those alone.

**Comment by daleharvey** [comment]
Thats awesome thanks a lot, just to clarify you expect the test-memleak to fail at the current time due to the still existing leaks right? if so then I am 👍 on getting this merged then we can work on fixing the rest and enabling the test in travis

**Comment by shimaore** [comment]
Yup, at this time the test for the http adapter is still failing (there is still a memleak somewhere). All the other tests should be successful.

**Comment by daleharvey** [comment]
Thats awesome, yeh I seen the same in the travis run :) sweet this is a great job, bugs like this are a nightmare to track down and test, very glad we have a set of tests in here now just to get them enabled :)

**Pull request merged**